### PR TITLE
Modernize varnames function

### PR DESCRIPTION
**Pull request by tgoodlet**
As per #19 this provides an update to `pluggy.varnames()`.
I've successfully run the unit tests across all interpreters on arch linux.
I'd like to test this across all dependent projects (`pytest`, `tox`, etc.) as well (and eventually get those tests integrated in a travis job for this project).

Note we'll need to move to `inspect.signature()` for py3.6...

### CODE REVIEW

File: pluggy.py [hunk 1]
```diff
@@ -627,50 +627,51 @@ def __repr__(self):
...
-        startindex = 1
-    else:
-        if not inspect.isfunction(func) and not inspect.ismethod(func):
-            try:
-                func = getattr(func, '__call__', func)
-            except Exception:
-                return ()
-        if startindex is None:
-            startindex = int(inspect.ismethod(func))
+    elif not inspect.isroutine(func):  # callable object?
```
Comment (RonnyPfannschmidt) [inline pluggy.py]: since what version of python does that method exist?

Comment (RonnyPfannschmidt) [inline pluggy.py]: never mind, i just verified its in 2.6

**Comment by hpk42** [comment]
I also like it -- @tgoodlet did you get a chance to test this against tox, pytest yet?  i guess doing develop installs from all of pluggy, pytest, tox and then running the respective test suites would go a long way.  Maybe pytest is not so imnportant actually because pluggy's test suite uses it and pytest will in turn use the latest pluggy and pytest is by far the heaviest user of pluggy i guess.  So i am merging this but i'd welcome if you watch this repository here in case related bugs turn up later.

**Comment by nicoddemus** [comment]
> ... pytest will in turn use the latest pluggy and pytest is by far the heaviest user of pluggy i guess

Actually we vendor `pluggy` in `pytest`, so pytest itself won't be using the local `pluggy`.

**Comment by hpk42** [comment]
On Fri, Nov 11, 2016 at 07:05 -0800, Bruno Oliveira wrote:
> > ... pytest will in turn use the latest pluggy and pytest is by far the heaviest user of pluggy i guess
> 
> Actually we vendor `pluggy` in `pytest`, so pytest itself won't be using the local `pluggy`.

good point.

**Comment by tgoodlet** [comment]
@hpk42 @nicoddemus @RonnyPfannschmidt that's one thing I tried to bring up before (pytest-dev/pytest#1637). I REALLY think in order to promote active development of pluggy `pytest` should be reversing it's import logic for the vendored versus locally installed `pluggy` version. If we start doing `pluggy` CI that includes running the test suites for dependent projects it makes it much easier to manage as well.

@nicoddemus had made the comment in that discussion:
>I think the point of vendoring is to isolate different pluggy installations in the same environment. For example, I right now both pytest and tox use pluggy in some vendored version, so different (and incompatible) versions of pluggy can coexist in the same virtual environment without problems.

I feel like this only applies to the case where you're running `pytest` to test `tox` (unless I'm missing something) since when running `tox` you should never be sharing `pluggy` with the underlying `virtualenv`s anyway. In other words there should never be a case where `pytest` and `tox` **must** share the same `pluggy` version under normal `tox` usage.

**Pull request merged**